url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/episodes/managers/FileSystemManager.java

```diff
@@ -0,0 +1,191 @@
+    }
+    
+    /**
+     * Attempts to add "path" to the list of local folders;
+     *  Ignores the path if invalid or could not be found
+     * 
+     * If path matches "@NAME/..", we assume it's a Windows
+     *  system and attempt to find an external harddrive
+     *  whose label matches NAME
+     * @param path
+     */
+    private static void addLocalFolder(List<Path> localFolders, String path) {
+        if (path.charAt(0) == '@') {
+            int end = path.indexOf( File.separatorChar );
+            if (end < 0)
+                return;
+            
+            String needle = path.substring(1, end);
+            String actualPath = path.substring(end+1);
+            
+            File[] roots = File.listRoots();
+            FileSystemView v = FileSystemView.getFileSystemView();
+            String label;
+            for (File f : roots) {
+                label = v.getSystemDisplayName(f);
+                if (label.startsWith(needle)) {
+                    localFolders.add( new Path(f.getAbsolutePath() + actualPath) );
+                    break;
+                }
+            }
+        } else {
+            // just add it
+            localFolders.add( new Path(path) );
+        }
+    }
+	   
+    protected static void loadFolderInfo(String folderInfo, List<Path> localFolders) {
+        // load the local folders list
+        if (folderInfo.contains(File.pathSeparator)) {
+            // it's definitely a list
+            String[] paths = folderInfo.split(File.pathSeparator);
+            addLocalFolders( localFolders, paths );
+        } else {
+            File cfgFile = new File(folderInfo);
+            if (!cfgFile.exists() && folderInfo.charAt(0) != '@') {
+                System.out.println(" - No folders provided; Failed.");
+                System.err.println("---------- DEBUG ----------");
+                System.err.println(" FolderInfo: " + folderInfo);
+                System.err.println(" file = " + cfgFile);
+                if (cfgFile != null) {
+                    System.err.println(" file.path = " + cfgFile.getAbsolutePath());
+                    System.err.println(" file.exists = " + cfgFile.exists());
+                    System.err.println(" file.canRead = " + cfgFile.canRead());
+                    System.err.println(" file.isDirectory = " + cfgFile.isDirectory());
+                }
+                System.err.println("---------- DEBUG ----------");
+                System.exit(1);
+            } 
+                
+            if (cfgFile.isFile()) {
+                // it's file; read line by line
+                ArrayList<String> folders = new ArrayList<String>();
+                try {
+                    BufferedReader reader = new BufferedReader(
+                            new InputStreamReader(
+                            new DataInputStream(
+                                    new FileInputStream(cfgFile.getCanonicalFile())
+                                    ))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *InputStreamReader*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *InputStreamReader()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).